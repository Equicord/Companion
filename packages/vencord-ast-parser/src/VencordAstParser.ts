import {
    AstParser,
    findObjectLiteralByKey,
    findParent,
    getImportName,
    getImportSource,
    Import,
    isDefaultImport,
    isInImportStatment,
    WithParent,
} from "@vencord-companion/ast-parser";
import { Cache, CacheGetter } from "@vencord-companion/shared/decorators";
import { Logger, NoopLogger } from "@vencord-companion/shared/Logger";

import { FindUse, FunctionNode, IFindType, IReplacement, PatchData, SourcePatch, StringNode, TestFind } from "./types";
import { tryParseRegularExpressionLiteral } from "./util";

import { readFile } from "node:fs/promises";
import { DeclarationDomain } from "ts-api-utils";
import {
    CallExpression,
    createPrinter,
    EmitHint,
    Expression,
    Identifier,
    isArrayLiteralExpression,
    isArrowFunction,
    isCallExpression,
    isFunctionExpression,
    isNamespaceImport,
    isObjectLiteralExpression,
    isPropertyAssignment,
    isRegularExpressionLiteral,
    isStringLiteral,
    isStringLiteralLike,
    isTemplateExpression,
    isVariableDeclaration,
    Node,
    ObjectLiteralExpression,
    ScriptTarget,
    transpileModule,
} from "typescript";


let logger: Logger = NoopLogger;

export function setLogger(newLogger: Logger) {
    logger = newLogger;
}

export class VencordAstParser extends AstParser {
    private readonly _path: string;

    public get path(): string {
        return this._path;
    }

    @CacheGetter()
    public get imports(): Map<Identifier, Import> {
        return this.listImports();
    }

    constructor(content: string, path: string) {
        super(content);
        this._path = path;
    }

    public static async fromPath(path: string) {
        return new VencordAstParser(await readFile(path, "utf-8"), path);
    }

    @Cache()
    private findDefinePlugin(): ObjectLiteralExpression | undefined {
        const define = [...this.imports.values()].find((x) => {
            if (!x.default)
                return;
            if (x.source !== "@utils/types")
                return;
            return true;
        });

        if (!define)
            return;

        const uses = this.vars.get(define.as)?.uses;

        if (!uses)
            return;

        const definePlugin = uses.find(({ location }) => {
            if (!isCallExpression(location.parent))
                return;
            return location.parent.arguments.length === 1 && isObjectLiteralExpression(location.parent.arguments[0]);
        });

        return (definePlugin?.location.parent as CallExpression).arguments[0] as ObjectLiteralExpression;
    }

    // TODO: work on files in the plugin folder but not the root plugin file
    public getPluginName(): string | null {
        const definePlugin = this.findDefinePlugin();

        if (!definePlugin)
            return null;

        const nameProp = findObjectLiteralByKey(definePlugin, "name");

        if (!nameProp || !isPropertyAssignment(nameProp) || !isStringLiteral(nameProp.initializer))
            return null;
        return nameProp.initializer.text;
    }

    @Cache()
    public getPatches(): SourcePatch[] {
        const definePlugin = this.findDefinePlugin();

        if (!definePlugin)
            return [];

        const patchesProp = findObjectLiteralByKey(definePlugin, "patches");

        if (!patchesProp || !isPropertyAssignment(patchesProp) || !isArrayLiteralExpression(patchesProp.initializer))
            return [];
        return patchesProp.initializer.elements
            .map((x, origIndex) => {
                if (!isObjectLiteralExpression(x))
                    return null;

                const res = this.parsePatch(x);

                if (!res)
                    return null;
                return {
                    ...res,
                    range: this.makeRangeFromAstNode(x.getChildAt(1)),
                    origIndex,
                };
            })
            .filter((x) => x !== null);
    }

    parseFind(patch: ObjectLiteralExpression): IFindType | null {
        const find = findObjectLiteralByKey(patch, "find");

        if (!find || !isPropertyAssignment(find))
            return null;
        if (!(isStringLiteral(find.initializer) || isRegularExpressionLiteral(find.initializer)))
            return null;

        return {
            findType: isStringLiteral(find.initializer) ? "string" : "regex",
            find: find.initializer.text,
        };
    }

    /**
     * Try to parse a string literal
     *
     * if it is a template literal, attempt to extract the string content by inlining variables
     */
    tryParseStringLiteral(node: Node): string | null {
        tryParse: if (isStringLiteralLike(node)) {
            return node.text;
            // resolve template literals if they are constant
        } else if (isTemplateExpression(node)) {
            const resolvedSpans = [] as string[];

            for (const span of node.templateSpans) {
                const spanExpr = span.expression;

                if (!this.isIdentifier(spanExpr)) {
                    logger.debug(`[VencordAstParser] Trying to parse template literal with non-identifier span: ${span.getText()}, FileName: ${span.getSourceFile().fileName}`);
                    break tryParse;
                }

                const usageInfo = this.getVarInfoFromUse(spanExpr);

                if (!usageInfo) {
                    break tryParse;
                } else if (usageInfo.declarations.length === 0) {
                    logger.trace(`[VencordAstParser] Could not resolve identifier ${spanExpr.text} to a variable declaration. Is it a global?`);
                    break tryParse;
                }

                const isConst = this.isConstDeclared(usageInfo);

                if (!isConst) {
                    break tryParse;
                }

                const decl = findParent(isConst[0], isVariableDeclaration);

                if (!decl) {
                    break tryParse;
                }

                const init = decl.initializer;

                if (!init) {
                    break tryParse;
                }

                const initValue = this.tryParseStringLiteral(init);

                // explicitly check for null to avoid empty string
                if (initValue == null) {
                    break tryParse;
                }

                resolvedSpans.push(initValue + span.literal.text);
            }

            return node.head.text + resolvedSpans.join("");
        }
        return null;
    }

    tryParseStringLiteralToStringNode(node: Node): StringNode | null {
        const str = this.tryParseStringLiteral(node);

        if (str == null)
            return null;

        return {
            type: "string",
            value: str,
        };
    }

    tryParseFunction(node: Node): FunctionNode | null {
        if (!isArrowFunction(node) && !isFunctionExpression(node))
            return null;

        const code = createPrinter()
            .printNode(EmitHint.Expression, node, node.getSourceFile());

        const res = transpileModule(code, {
            compilerOptions: {
                target: ScriptTarget.ESNext,
                strict: true,
            },
        });

        if (res.diagnostics && res.diagnostics.length > 0)
            return null;

        return {
            type: "function",
            value: res.outputText,
        };
    }

    parseReplace(node: Expression) {
        return this.tryParseStringLiteralToStringNode(node) ?? this.tryParseFunction(node);
    }

    parseMatch(node: Expression) {
        return this.tryParseStringLiteralToStringNode(node) ?? tryParseRegularExpressionLiteral(node);
    }

    parseReplacement(patch: ObjectLiteralExpression): IReplacement[] | null {
        const replacementObj = findObjectLiteralByKey(patch, "replacement");

        if (!replacementObj || !isPropertyAssignment(replacementObj))
            return null;

        const replacement = replacementObj.initializer;
        const replacements = isArrayLiteralExpression(replacement) ? replacement.elements : [replacement];

        if (!replacements.every(isObjectLiteralExpression))
            return null;

        const replacementValues = (replacements as ObjectLiteralExpression[]).map((r: ObjectLiteralExpression) => {
            const match = findObjectLiteralByKey(r, "match");
            const replace = findObjectLiteralByKey(r, "replace");

            if (!replace || !isPropertyAssignment(replace) || !match || !isPropertyAssignment(match))
                return null;

            const matchValue = this.parseMatch(match.initializer);

            if (!matchValue)
                return null;

            const replaceValue = this.parseReplace(replace.initializer);

            if (replaceValue == null)
                return null;

            return {
                match: matchValue,
                replace: replaceValue,
            };
        })
            .filter((x) => x != null);

        return replacementValues.length > 0 ? replacementValues : null;
    }

    parsePatch(patch: ObjectLiteralExpression): PatchData | null {
        const find = this.parseFind(patch);
        const replacement = this.parseReplacement(patch);

        if (!replacement || !find)
            return null;

        return {
            ...find,
            replacement,
        };
    }


    /**
     * @returns true if this file is the entry point (if it has a definePlugin call)
     */
    public isRootPluginFile(): boolean {
        return !!this.findDefinePlugin();
    }

    @Cache()
    public getFinds(): FindUse[] {
        return this.getFindUses()
            .map<FindUse | false>((x) => {
                const call = x.parent;

                if (call.arguments.length === 0)
                    return false;

                const args = call.arguments.map((x) => {
                    return this.tryParseStringLiteralToStringNode(x)
                      ?? tryParseRegularExpressionLiteral(x)
                      ?? this.tryParseFunction(x);
                });

                const range = this.makeRangeFromAstNode(call);

                return {
                    range,
                    use: {
                        type: "testFind",
                        data: {
                            type: x.getText() as TestFind["data"]["type"],
                            args: args.filter((x) => x != null),
                        },
                    },
                };
            })
            .filter((x) => x !== false);
    }

    @Cache()
    private getFindUses(): WithParent<Identifier, CallExpression>[] {
        const imports = [...this.imports.entries()].flatMap(([k, v]) => {
            if (v.source !== "@webpack")
                return [];

            const origName = (v.orig ?? v.as).getText();

            if (!origName.startsWith("find"))
                return [];
            return k;
        });

        const toRet: WithParent<Identifier, CallExpression>[] = [];

        for (const i of imports) {
            const uses = this.vars.get(i)?.uses;

            if (!uses)
                continue;
            for (const { location } of uses) {
                if (!isCallExpression(location.parent))
                    continue;
                toRet.push(location as WithParent<Identifier, CallExpression>);
            }
        }
        return toRet;
    }

    /**
     * returns the import if this identifier is imported
     */
    private isIdentifierImported(i: Identifier): Import | undefined {
        const { declarations, domain } = this.vars.get(i) ?? {};

        if (!declarations || declarations.length === 0)
            return;
        if (!(domain! & DeclarationDomain.Import))
            return;

        const source = declarations.flatMap((x) => {
            if (!isInImportStatment(x))
                return [];
            return x;
        });

        if (source.length !== 1)
            return;

        const [importIdent] = source;

        return {
            default: isDefaultImport(importIdent),
            namespace: isNamespaceImport(importIdent),
            ...getImportName(importIdent),
            source: getImportSource(importIdent),
        };
    }

    private listImports(): Map<Identifier, Import> {
        return new Map([...this.vars.entries()].flatMap(([k]) => {
            const ret = this.isIdentifierImported(k);

            return ret ? [[k, ret]] : [];
        }));
    }
}
