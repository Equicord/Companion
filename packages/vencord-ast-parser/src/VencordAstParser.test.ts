import { VencordAstParser } from "./VencordAstParser";

import { ChildProcess, exec } from "child_process";
import { Dirent, PathLike, readFileSync } from "fs";
import { readdir, readFile, stat } from "fs/promises";
import { join } from "path";

import { assert, describe, expect, it } from "vitest";

const __dirname = import.meta.dirname;
const VENCORD_DIR = join(__dirname, "__test__", ".vencord-source");
const VENCORD_REV = "8807564053c7b4cc05c763e2dc7171f5d61e39c7";

function parserFor(path: string): VencordAstParser {
    path = join(__dirname, "__test__", path);
    return new VencordAstParser(readFileSync(path, "utf-8"), path);
}

describe("VencordAstParser", async function () {
    await ensureVencordDownloaded();

    // collect all plugin paths
    const pluginParsers = await Promise.all((await collectPluginPaths()).map(async (path) => new VencordAstParser(await readFile(path, "utf-8"), path)));

    describe("getPluginName", function () {
        it("parses all plugin names to non-null values", function () {
            for (const parser of pluginParsers) {
                const name = parser.getPluginName();

                expect(name, `Parsing plugin name failed for plugin at path ${parser.path}`).to.be.a("string");
            }
        });
        it("parses all plugin names correctly", function () {
            const names = pluginParsers.map((parser) => parser.getPluginName());

            expect(names)
                .toMatchSnapshot();
        });
        it.skip("gets the correct plugin name for a weird plugin", function () {
            const parser = parserFor("pluginImports.ts");

            expect(parser.getPluginName()).to.equal("2");
        });
    });
});

function waitForProcess(process: ChildProcess): Promise<void> {
    return new Promise<void>((res, rej) => {
        process.on("exit", (code) => {
            if (code === 0) {
                res();
            } else {
                rej(new Error(`Child process exited with code: ${code}`));
            }
        });
    });
}

async function ensureVencordDownloaded() {
    if (await exists(VENCORD_DIR) && await isDirectory(VENCORD_DIR)) {
        return;
    }


    await waitForProcess(exec(`git clone https://github.com/vendicated/vencord.git ${VENCORD_DIR}`));
    await waitForProcess(exec(`git checkout --detach ${VENCORD_REV}`, {
        cwd: VENCORD_DIR,
    }));
}

async function resolvePluginEntrypoint(pluginEntry: Dirent): Promise<string> {
    const pluginEntryPath = join(pluginEntry.parentPath, pluginEntry.name);

    if (pluginEntry.isFile()) {
        return pluginEntryPath;
    }

    let path = join(pluginEntryPath, "index.ts");

    if (await exists(path)) {
        return path;
    } else if (await exists(path = join(pluginEntryPath, "index.tsx"))) {
        return path;
    }
    throw new Error("No valid entry point found");
}

async function collectPluginPaths(): Promise<string[]> {
    const basePluginDir = join(VENCORD_DIR, "src", "plugins");
    const pluginDirs = [basePluginDir, join(basePluginDir, "_api"), join(basePluginDir, "_core")];

    for (const dir of pluginDirs) {
        assert(await isDirectory(dir));
    }

    const pluginFiles = await Promise.all(pluginDirs.map((dir) => readdir(dir, { withFileTypes: true })));

    return await Promise.all(pluginFiles
        .flat()
        .filter((dir) => {
            // don't match index.ts at root of plugin tree, it's not a plugin
            return dir.name[0] !== "_" && dir.name !== "index.ts";
        })
        .map(resolvePluginEntrypoint));
}

async function exists(path: PathLike) {
    try {
        return !!await stat(path);
    } catch {
        return false;
    }
}
/**
 * **PATH MUST EXIST**
 *
 * \@throws if the path doesnt exist
 *
 * use {@link exists} to check if it exists
 */
async function isDirectory(path: PathLike) {
    return (await stat(path)).isDirectory();
}
