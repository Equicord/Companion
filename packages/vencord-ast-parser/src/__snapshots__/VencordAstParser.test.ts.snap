// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VencordAstParser > getPluginName > parses all plugin names correctly 1`] = `
[
  "AccountPanelServerProfile",
  "AlwaysAnimate",
  "AlwaysExpandRoles",
  "AlwaysTrust",
  "AnonymiseFileNames",
  "AppleMusicRichPresence",
  "WebRichPresence (arRPC)",
  "BetterFolders",
  "BetterGifAltText",
  "BetterGifPicker",
  "BetterNotesBox",
  "BetterRoleContext",
  "BetterRoleDot",
  "BetterSessions",
  "BetterSettings",
  "BetterUploadButton",
  "BiggerStreamPreview",
  "BlurNSFW",
  "CallTimer",
  "ClearURLs",
  "ClientTheme",
  "ColorSighted",
  "ConsoleJanitor",
  "ConsoleShortcuts",
  "CopyEmojiMarkdown",
  "CopyFileContents",
  "CopyStickerLinks",
  "CopyUserURLs",
  "CrashHandler",
  "CtrlEnterSend",
  "CustomIdle",
  "CustomRPC",
  "Dearrow",
  "Decor",
  "DevCompanion",
  "DisableCallIdle",
  "DontRoundMyTimestamps",
  "Experiments",
  "ExpressionCloner",
  "F8Break",
  "FakeNitro",
  "FakeProfileThemes",
  "FavoriteEmojiFirst",
  "FavoriteGifSearch",
  "FixCodeblockGap",
  "FixImagesQuality",
  "FixSpotifyEmbeds",
  "FixYoutubeEmbeds",
  "ForceOwnerCrown",
  "FriendInvites",
  "FriendsSince",
  "FullSearchContext",
  "FullUserInChatbox",
  "GameActivityToggle",
  "GifPaste",
  "GreetStickerPicker",
  "HideMedia",
  "iLoveSpam",
  "IgnoreActivities",
  "ImageFilename",
  "ImageLink",
  "ImageZoom",
  "ImplicitRelationships",
  "InvisibleChat",
  "IrcColors",
  "KeepCurrentChannel",
  "LastFMRichPresence",
  "LoadingQuotes",
  "MemberCount",
  "MentionAvatars",
  "MessageClickActions",
  "MessageLatency",
  "MessageLinkEmbeds",
  "MessageLogger",
  "MessageTags",
  "MutualGroupDMs",
  "NewGuildSettings",
  "NoBlockedMessages",
  "DisableDeepLinks",
  "NoDevtoolsWarning",
  "NoF1",
  "NoMaskedUrlPaste",
  "NoMosaic",
  "NoOnboardingDelay",
  "NoPendingCount",
  "NoProfileThemes",
  "NoReplyMention",
  "NoServerEmojis",
  "NoSystemBadge",
  "NoTypingAnimation",
  "NoUnblockToJump",
  "NormalizeMessageLinks",
  "NotificationVolume",
  "OnePingPerDM",
  "oneko",
  "OpenInApp",
  "OverrideForumDefaults",
  "PauseInvitesForever",
  "PermissionFreeWill",
  "PermissionsViewer",
  "petpet",
  "PictureInPicture",
  "PinDMs",
  "PlainFolderIcon",
  "PlatformIndicators",
  "PreviewMessage",
  "QuickMention",
  "QuickReply",
  "ReactErrorDecoder",
  "ReadAllNotificationsButton",
  "RelationshipNotifier",
  "ReplaceGoogleSearch",
  "ReplyTimestamp",
  "RevealAllSpoilers",
  "ReverseImageSearch",
  "ReviewDB",
  "RoleColorEverywhere",
  "SecretRingToneEnabler",
  "Summaries",
  "SendTimestamps",
  "ServerInfo",
  "ServerListIndicators",
  "ShikiCodeblocks",
  "ShowAllMessageButtons",
  "ShowConnections",
  "ShowHiddenChannels",
  "ShowHiddenThings",
  "ShowMeYourName",
  "ShowTimeoutDuration",
  "SilentMessageToggle",
  "SilentTyping",
  "SortFriendRequests",
  "SpotifyControls",
  "SpotifyCrack",
  "SpotifyShareCommands",
  "StartupTimings",
  "StickerPaste",
  "StreamerModeOnStream",
  "SuperReactionTweaks",
  "TextReplace",
  "ThemeAttributes",
  "Translate",
  "TypingIndicator",
  "TypingTweaks",
  "Unindent",
  "UnlockedAvatarZoom",
  "UnsuppressEmbeds",
  "UserMessagesPronouns",
  "UserVoiceShow",
  "USRBG",
  "ValidReply",
  "ValidUser",
  "VoiceChatDoubleClick",
  "VcNarrator",
  "VencordToolbox",
  "ViewIcons",
  "ViewRaw",
  "VoiceDownload",
  "VoiceMessages",
  "VolumeBooster",
  "WebContextMenus",
  "WebKeybinds",
  "WebScreenShareFixes",
  "WhoReacted",
  "XSOverlay",
  "YoutubeAdblock",
  "BadgeAPI",
  "ChatInputButtonAPI",
  "CommandsAPI",
  "ContextMenuAPI",
  "DynamicImageModalAPI",
  "MemberListDecoratorsAPI",
  "MenuItemDemanglerAPI",
  "MessageAccessoriesAPI",
  "MessageDecorationsAPI",
  "MessageEventsAPI",
  "MessagePopoverAPI",
  "MessageUpdaterAPI",
  "NoticesAPI",
  "ServerListAPI",
  "UserSettingsAPI",
  "NoTrack",
  "Settings",
  "SupportHelper",
]
`;
